File: hangar/environment.js
/** Generic Impossible Hangar transition kernel. Configurations supply rules; this
 * engine deliberately contains no winning sequence or puzzle-specific policy. */
export class HangarEnvironment {
  #configuration;
  #state;
  #step = 0;
  #history = [];

  constructor(configuration) {
    validateConfiguration(configuration);
    this.#configuration = structuredClone(configuration);
    this.#state = structuredClone(configuration.initialState);
  }

  observe() {
    const visible = {};
    for (const [name, sensor] of Object.entries(this.#configuration.sensors)) {
      if (!sensor.availableWhen || conditionsMatch(this.#state, sensor.availableWhen)) {
        visible[name] = sensor.fields.reduce((result, field) => ({ ...result, [field]: this.#state[field] }), {});
      }
    }
    return { step: this.#step, observations: visible, resources: structuredClone(this.#state.resources ?? {}) };
  }

  act(actionName) {
    const action = this.#configuration.actions[actionName];
    if (!action) return this.#record(actionName, false, 'unknown_action', {});
    if (!conditionsMatch(this.#state, action.requires ?? {})) return this.#record(actionName, false, 'requirements_not_met', {});
    if (!resourcesAvailable(this.#state.resources ?? {}, action.cost ?? {})) return this.#record(actionName, false, 'insufficient_resources', {});
    applyCost(this.#state.resources ?? {}, action.cost ?? {});
    const before = structuredClone(this.#state);
    Object.assign(this.#state, structuredClone(action.effects ?? {}));
    this.#step += 1;
    return this.#record(actionName, true, 'applied', changedFields(before, this.#state));
  }

  isMissionSuccessful() {
    return conditionsMatch(this.#state, this.#configuration.success);
  }

  history() { return structuredClone(this.#history); }

  #record(action, success, result, changes) {
    const event = { step: this.#step, action, success, result, changes };
    this.#history.push(event);
    return structuredClone(event);
  }
}

function conditionsMatch(state, conditions) { return Object.entries(conditions).every(([key, expected]) => Object.is(state[key], expected)); }
function resourcesAvailable(resources, cost) { return Object.entries(cost).every(([key, amount]) => Number(resources[key] ?? 0) >= Number(amount)); }
function applyCost(resources, cost) { for (const [key, amount] of Object.entries(cost)) resources[key] = Number(resources[key] ?? 0) - Number(amount); }
function changedFields(before, after) { const changes = {}; for (const key of new Set([...Object.keys(before), ...Object.keys(after)])) if (JSON.stringify(before[key]) !== JSON.stringify(after[key])) changes[key] = structuredClone(after[key]); return changes; }
function validateConfiguration(configuration) {
  if (!configuration || typeof configuration !== 'object') throw new TypeError('configuration is required');
  for (const field of ['initialState', 'sensors', 'actions', 'success']) if (!configuration[field] || typeof configuration[field] !== 'object') throw new TypeError(`${field} is required`);
}

File: hangar/environment.py
from copy import deepcopy
class HangarEnvironment:
 def __init__(self,config):
  for key in ('initial_state','sensors','actions','success'):
   if not isinstance(config.get(key),dict): raise TypeError(f'{key} is required')
  self.config=deepcopy(config);self.state=deepcopy(config['initial_state']);self.initial_resources=deepcopy(self.state.get('resources',{}));self.step=0;self.actions=[];self.observations=[]
 def observe(self):
  visible={name:{f:self.state.get(f) for f in sensor['fields']} for name,sensor in self.config['sensors'].items() if not sensor.get('available_when') or self._matches(sensor['available_when'])};result={'step':self.step,'observations':visible,'resources':deepcopy(self.state.get('resources',{}))};self.observations.append(deepcopy(result));return result
 def act(self,name):
  action=self.config['actions'].get(name)
  if not action:return self._record(name,False,'unknown_action',{})
  if not self._matches(action.get('requires',{})):return self._record(name,False,'requirements_not_met',{})
  if any(self.state.get('resources',{}).get(k,0)<v for k,v in action.get('cost',{}).items()):return self._record(name,False,'insufficient_resources',{})
  before=deepcopy(self.state)
  for k,v in action.get('cost',{}).items():self.state['resources'][k]-=v
  self.state.update(deepcopy(action.get('effects',{})));self.step+=1;changes={k:deepcopy(v) for k,v in self.state.items() if before.get(k)!=v};return self._record(name,True,'applied',changes)
 def _matches(self,conditions):return all(self.state.get(k)==v for k,v in conditions.items())
 def _record(self,name,success,result,changes):event={'step':self.step,'action':name,'success':success,'result':result,'changes':changes};self.actions.append(event);return deepcopy(event)
 def available_actions(self):return list(self.config['actions'])
 def mission_successful(self):return self._matches(self.config['success'])
 def episode(self):return {'actions':deepcopy(self.actions),'observations':deepcopy(self.observations),'initial_resources':deepcopy(self.initial_resources),'final_resources':deepcopy(self.state.get('resources',{})),'mission_success':self.mission_successful()}
